### DISCUSSION TITLE
protected evaluate method does not allow to use return values e.g. in tests.

### DISCUSSION THREAD

**Discussion by martin-grofcik** [spring-projects/spring-shell#897] [Q&A]
I have migrated my project from spring-shell 2.x to 3.1.4. I was able to solve all differences.
The one, which I was not able to solve in the clear way 
```
	/**
	 * Evaluate a single "line" of input from the user by trying to map words to a command and
	 * arguments.
	 *
	 * <p>
	 * This method does not throw exceptions, it catches them and returns them as a regular
	 * result
	 * </p>
	 */
	protected Object evaluate(Input input) {
```

I used evaluate method in 2.x tests a lot:
```
        ObjectNode startProcessResult = (ObjectNode) shell.evaluate(
                () -> "ex POST /process-api/runtime/process-instances {\"processDefinitionKey\":\"oneTaskProcess\",\"businessKey\":\"myBusinessKey\"}"
        );
        String processId = startProcessResult.get("id").asText();
```
I used shell wrapper which allows to execute evaluate method and provide the object as output.
Is there any clear way how to have the same functionality in 3.1.x?

**Comment by jvalkeal** [comment]
There's a new testing system https://docs.spring.io/spring-shell/reference/testing/basics.html which might help you. Though it's more for e2e as result is what's visible in an emulated test screen. There's also `org.springframework.test.util.ReflectionTestUtils` which many spring projects use in tests to access non-public methods and fields.

Is there a specific reason you want to test through that evaluate method?

**Comment by martin-grofcik** [comment]
Thank you for your answer. 

Exactly as you said e2e, assertion based on the displayed text.
> Is there a specific reason you want to test through that evaluate method?
as I wrote:
```
        ObjectNode startProcessResult = (ObjectNode) shell.evaluate(
                () -> "ex POST /process-api/runtime/process-instances {\"processDefinitionKey\":\"oneTaskProcess\",\"businessKey\":\"myBusinessKey\"}"
        );
        String processId = startProcessResult.get("id").asText();
```
I need to you output of the evaluate method in the following calls. In this case `processId` is outcome written in json. To parse `session.screen().getLines()` and find processId is not straights forward.

**Comment by kdziuba-opgk** [comment]
I got similar issue. In my case I need to evaluate strings provided by the user. It's a little disappointing there is no easy access to the command output.
